src/game_launcher.cpp

```diff
@@ -834,21 +834,15 @@ void game_launcher::start_wesnothd()
 		filesystem::write_file(config, filesystem::read_file(filesystem::get_wml_location("lan_server.cfg")));
 	}
 
+	LOG_GENERAL << "Starting wesnothd";
+	try
+	{
 #ifndef _WIN32
-	std::string command = "\"" + wesnothd_program +"\" -c \"" + config + "\" -d -t 2 -T 5";
+		bp::spawn(wesnothd_program, "-c", "\""+config+"\"", "-d");
```

Comment:
Don't surround the config with quotes. That'll almost certainly fail.

Reply:
Also need to call `search_path` since `wesnothd_program` isn't always the full path.

Reply:
Using `search_path` results in it not finding the wesnothd executable, whether it's using the full path or not.

Reply:
Ok, that's weird. That is pretty much the example from the [documentation](https://www.boost.org/doc/libs/1_82_0/doc/html/boost_process/tutorial.html#boost_process.tutorial.launch_mode).
Maybe `search_path` doesn't like it when you give it a path already.

Well, if it works without `search_path` even when it's just wesnothd then that's fine.

Reply:
To clarify (with both `wesnoth` and `wesnothd` in my local repo's root folder):
* If I use `search_path`, then with either just `wesnothd` or the full path it fails to find the executable.
* If I don't use `search_path`, then with either just `wesnothd` or the full path it does find the executable.
* I didn't test it with `wesnothd` being somewhere else that would be found from searching the PATH locations.

Reply:
Apparently `search_path` only works if you pass it _just_ the program name. Glancing over the code, it appears that maybe `get_program_invocation` is doing the same thing that `search_path` would? I don't know for certain though.

Reply:
I tried passing just the program name by doing `wesnothd_program = filesystem::base_name(wesnothd_program);` and it still failed if I used `search_path`.

Reply:
Weird… but still, am I correct about what `get_program_invocation` does? If so, then it's correct not to use `search_path` anyway, which I suppose is a good thing if `search_path` just doesn't work.

Reply:
The main point being that it returns an absolute path, regardless of the method.

So given that, using `search_path` would be wrong even if it functions.

Reply:
You can definitely start wesnothd just by its name if that is the question.

Reply:
Which does not work anymore now.

Reply:
If that functionality is important, I suppose you could add something like:

```cpp
if(wesnothd_program.find('/') == std::string::npos) {
    wesnothd_program = bp::search_path(wesnothd_program);
}
```

Reply:
Would be good to not require people to pick the wesnothd binary if we can avoid it.

Reply:
I guess the only problem with that would be, if the binary is in PATH, what are the chances that it's the wrong binary (a different version)?

Reply:
That would've already been a risk though, right?

Reply:
Yes.

Reply:
So I installed the wesnoth-server package (installing the client doesn't also install the server executable), made a copy of `wesnothd-1.16` named as just `wesnothd` so that `which wesnothd` returns `/usr/games/wesnothd`. `bp::search_path` however still does not find anything - `bp::search_path(filesystem::base_name(wesnothd_program)).string()` returns an empty string.

Reply:
Also I'm confused over how the previous implementation would've searched the path? `filesystem::get_program_invocation` certainly wouldn't have done that, or given a value that would've resulted in that being done by `std::system`.

Reply:
It's because `std::system` invokes the shell, which of _course_ searches the path when running a command by name.

Reply:
It seems fairly irrelevant what the preference file contains. If people put nonsense in there than they can also fix it and put the right thing in there.

The default is determined by argv[0] so it's up to the caller what it puts there. If you call wesnoth with the full path there is probably a good chance the caller also puts the full path in argv[0] and if you just call wesnoth then there is a good chance it'll also just contain wesnoth.

Reply:
Since using argv[0] is not particularly great and we already have `get_exe_dir` which does a better job at finding the wesnoth binary dir (at least on windows and anything with /proc like linux) and is used for the starting point for the wesnothd path picking dialog we could use that in `get_program_invocation` as well. Though it's not quite ideal that it falls back to `get_cwd` a fallback to PATH search (and then to cwd) would be better still. (And better handling for OSes without /proc like macos.)

Otherwise the code CelticMinstrel suggested is pretty much what a shell does.

Reply:
> It's because std::system invokes the shell, which of course searches the path when running a command by name.

Right, but `filesystem::get_program_invocation` will never give you just the program name, it will always give you the full path to where it guesses wesnothd will be. In which case in practice that would never have resulted in a PATH search because it would always have used a full path to the wesnothd executable. So adding functionality to try and find wesnothd via the PATH would be new functionality.

Though I suppose it's not especially relevant either way until `search_path` returns  an actual path to use.

Reply:
But the other branch, that looks up the path from the preferences, theoretically could return just the program name… though only if the user manually edited the preferences file, I think.

Reply:
`get_program_invocation` will return just the program name as explained above. No need to edit preferences.

Reply:
https://github.com/wesnoth/wesnoth/blob/a6883aabe0845f219cbb813fda3c3450086a6415/src/wesnoth.cpp#L754
`args[0]` there is the same as `argv[0]` which is whatever the caller put there. That might be the full path to wesnoth (relative or absolute) or it might just be wesnoth (or really anything the caller wants but usually one of those options).

These are the log messages I get from the flatpak build:
> 20230823 10:13:54 warning general: Failed to start server wesnothd:
execve failed: No such file or directory

Reply:
So the changes would be to:
* use `filesystem::get_exe_dir()` instead of `game_config::wesnoth_program_dir`.
* update `filesystem::get_exe_dir()` to check `bp::search_path()` results before falling back to `get_cwd()`?

Reply:
Yeah, that sounds good.

Probably still need to use `get_program_invocation` to get the right program name for `bp::search_path()` but I'd remove the `game_config::wesnoth_program_dir` part.